# Cardiac Patient Monitoring System

## Milestone 1 — Understanding the Data

### Introduction

This project begins with the Heart Disease Cleveland dataset, a collection of patient records containing cardiac and clinical features that may be related to the presence of heart disease.

Rather than jumping directly into model training, I want to first understand the patients represented in this dataset and the quality of the information available about them.

The dataset contains **303 patient records**, described by **13 input features**, with the **`target`** column representing the outcome that will later be predicted.

Before asking a machine learning model to learn from these records, I need to answer a more basic question:

**Is the data itself ready to be trusted?**

In this milestone, I will inspect the structure of the dataset, understand what each feature represents, examine the target variable, and look for missing, duplicated, or inconsistent values that could affect the analysis later.

### Environment

This project is developed in Jupyter Notebook using Python with Pandas, NumPy, Matplotlib, and Scikit-learn.

In [23]:
import sys
import pandas as pd
import numpy as np

print("Python version:", sys.version.split()[0])
print("Pandas version:", pd.__version__)
print("NumPy version:", np.__version__)

Python version: 3.12.13
Pandas version: 3.0.3
NumPy version: 2.5.1


### Dataset Source

The dataset used in this project is the **Heart Disease Cleveland dataset**, downloaded from Kaggle for educational machine-learning analysis.

## 1. Loading the Dataset

The Cleveland Heart Disease dataset is the starting point of this analysis.

Before changing or cleaning anything, I want to see the data exactly as it was provided.

In [14]:
# Import Pandas for working with tabular data
import pandas as pd

# Load the Cleveland Heart Disease dataset
df = pd.read_csv("../data/Heart_disease_cleveland_new.csv")

# Display the first five patient records
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,0,145,233,1,2,150,0,2.3,2,0,2,0
1,67,1,3,160,286,0,2,108,1,1.5,1,3,1,1
2,67,1,3,120,229,0,2,129,1,2.6,1,2,3,1
3,37,1,2,130,250,0,0,187,0,3.5,2,0,1,0
4,41,0,1,130,204,0,2,172,0,1.4,0,0,1,0


### Observation

The first rows show that the dataset contains both numerical measurements and features represented using numeric codes.

The `target` column is also included and will later be used as the outcome for the classification task.

## 2. How Large Is the Dataset?

Now that I have seen what a patient record looks like, the next step is to understand the overall size of the dataset.

Knowing the number of patient records and columns gives a clearer picture of the amount of data available for the analysis.

In [15]:
# Check the number of rows and columns
df.shape

(303, 14)

### Observation

The dataset contains **303 patient records and 14 columns**.

Among these columns, **13 are input features** describing different patient characteristics, while the remaining column, **`target`**, represents the outcome that will later be used for classification.

This is a relatively small dataset, which makes careful evaluation especially important when the modeling stage begins.

## 3. What Information Does Each Patient Record Contain?

Knowing the size of the dataset is only part of the picture. The next step is to understand what information is stored for each patient and what each column represents.

In [16]:
# Display all column names
df.columns.tolist()

['age',
 'sex',
 'cp',
 'trestbps',
 'chol',
 'fbs',
 'restecg',
 'thalach',
 'exang',
 'oldpeak',
 'slope',
 'ca',
 'thal',
 'target']

### Observation

The dataset contains 13 patient features and one target column.

Some columns, such as `age`, `chol`, and `thalach`, contain direct numerical measurements, while others, such as `cp`, `restecg`, and `slope`, use numbers to represent categories.

## 4. Data Dictionary

The column names are short, so before continuing I want to define what each one represents.

| Feature | Description |
|---|---|
| `age` | Age of the patient in years |
| `sex` | Sex of the patient: 0 = female, 1 = male |
| `cp` | Chest pain type: 0 = typical angina, 1 = atypical angina, 2 = non-anginal pain, 3 = asymptomatic |
| `trestbps` | Resting blood pressure in mm Hg |
| `chol` | Serum cholesterol in mg/dl |
| `fbs` | Fasting blood sugar > 120 mg/dl: 0 = false, 1 = true |
| `restecg` | Resting ECG result: 0 = normal, 1 = ST-T abnormality, 2 = left ventricular hypertrophy |
| `thalach` | Maximum heart rate achieved |
| `exang` | Exercise-induced angina: 0 = no, 1 = yes |
| `oldpeak` | ST depression induced by exercise relative to rest |
| `slope` | Slope of the peak exercise ST segment: 0 = upsloping, 1 = flat, 2 = downsloping |
| `ca` | Number of major vessels |
| `thal` | Thal result: 1 = normal, 2 = fixed defect, 3 = reversible defect |
| `target` | Heart disease status: 0 = absence, 1 = presence |

## 5. Checking the Dataset Structure

Before cleaning the data, I want to check the data types and make sure all columns contain the expected number of values.

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


### Observation

All 14 columns contain 303 non-null values.

Most columns are stored as integers, while `oldpeak` is stored as a float because it contains decimal values.

At this point, there are no missing values detected by Pandas, but I will check this more directly in the next step.

## 6. Are there any missing values ? 
Missing values can affect both analysis and model , so i want to check whether any feature contains incomplete patient record 

In [18]:
# count missing values in each columns 
df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

### Observation 

All columns have 0 missing values.
This means the dataset is complete 

## 7. Are There Any Duplicate Records ? 

Even if the dataset is complete , repeated rows can still affect the analysis 
so i will check whether any recored appear more than once

In [19]:
# Count duplicate rows
df.duplicated().sum()

np.int64(0)

there is no duplicate rows so no rows need to be removed for duplication 

## 8. Are the Values Reasonable?

The dataset has no missing or duplicate records, but I still need to check whether the values themselves look valid.

I will inspect the unique values of the coded features to make sure they match the expected categories.

In [20]:
# Check unique values in coded categorical features 
categorical_col =[
    "sex",
    "cp",
    "fbs",

"restecg",
"exang",
"slope",
"ca",
"thal",
"target"
]
for columns in categorical_col:
    print(f"{columns}:{sorted(df[columns].unique())}")

sex:[np.int64(0), np.int64(1)]
cp:[np.int64(0), np.int64(1), np.int64(2), np.int64(3)]
fbs:[np.int64(0), np.int64(1)]
restecg:[np.int64(0), np.int64(1), np.int64(2)]
exang:[np.int64(0), np.int64(1)]
slope:[np.int64(0), np.int64(1), np.int64(2)]
ca:[np.int64(0), np.int64(1), np.int64(2), np.int64(3)]
thal:[np.int64(1), np.int64(2), np.int64(3)]
target:[np.int64(0), np.int64(1)]


### Observation

The coded features contain the expected category values.

Binary features such as `sex`, `fbs`, `exang`, and `target` contain only 0 and 1, while the other coded features contain a small set of valid category codes.

No unexpected category values were found.

## 9. Checking Numerical Ranges

The coded features look consistent, so next I want to inspect the numerical measurements and check whether their ranges look reasonable.

In [21]:
# View summary statistics for numerical features
df[["age", "trestbps", "chol", "thalach", "oldpeak"]].describe()

,age,trestbps,chol,thalach,oldpeak
count,303.000000,303.000000,303.000000,303.000000,303.000000
mean,54.438944,131.689769,246.693069,149.607261,1.039604
std,9.038662,17.599748,51.776918,22.875003,1.161075
min,29.000000,94.000000,126.000000,71.000000,0.000000
25%,48.000000,120.000000,211.000000,133.500000,0.000000
50%,56.000000,130.000000,241.000000,153.000000,0.800000
75%,61.000000,140.000000,275.000000,166.000000,1.600000
max,77.000000,200.000000,564.000000,202.000000,6.200000


### Observation

The numerical features show reasonable ranges overall.

Patient ages range from **29 to 77 years**, while resting blood pressure ranges from **94 to 200 mm Hg**.

The cholesterol values range from **126 to 564 mg/dl**, and the maximum heart rate ranges from **71 to 202 bpm**.

Some values, especially the maximum cholesterol and blood pressure values, are noticeably higher than the typical range of the dataset. I will keep them for now and examine them more closely during the EDA stage instead of removing them without evidence.

## 10. Initial Cleaning Decision

After checking missing values, duplicates, category codes, and numerical ranges, I can now decide whether the dataset needs any immediate cleaning before moving to EDA.

### Cleaning Summary

The dataset did not have any missing values or duplicate rows.

The category values also looked correct, and the numerical values were reasonable overall. Some high values appeared, but I decided to keep them and check them later during EDA.

At this stage, I did not remove or change any records because there was no clear reason to do that.

In [22]:
# Save the checked dataset for the next stages
df.to_csv("../data/heart_cleveland_checked.csv", index=False)

## Milestone 1 Takeaway

The dataset has **303 patient records, 13 input features, and one target column**.

After checking the data, I found no missing values or duplicate rows, and the category values were also correct.

The numerical values looked reasonable overall, so I did not remove any records at this stage.

Now the dataset is ready for the next step, where I will start exploring the distributions and relationships between the features.